### DISCUSSION TITLE
Clip / slice / interact with widgets over a whole scene ( all meshes/actors)

### DISCUSSION THREAD

**Discussion by T4mmi** [pyvista/pyvista#5282] [Q&A]
Hi, I'm looking for a way to clip a whole scene in a `pyvista.Plotter()` using a widget.  
The [`add_mesh_clip_box()`](https://docs.pyvista.org/version/stable/api/plotting/_autosummary/pyvista.plotting.widgets.WidgetHelper.add_mesh_clip_box.html#pyvista.plotting.widgets.WidgetHelper.add_mesh_clip_box) (and others alike) seems to be the way-to-go, but it requires to define a single mesh ...  
Yeah I'm aware I could combine a `pyvista.MultiBlock` ... but I've got a complex interactive scene with colors, transparency and texture applied individually to >100 objects/meshes.  

Is there any way to send at least a list of independent meshes to such filters ? or does anyone has any idea how to filter the display to ignore features outised the clipper (I don't mind if it's a box, a plane or even a line with vertical clipping) ?

**Comment by callaway-chris-yang** [comment]
Hi, may I ask if you have a solution for that?

**Reply by T4mmi** [reply]
Hi, I did not actually ... so my (dirty and inefficient) workaround is to loop over all actors and create a clip of each individually 😞 

As said its a dirty fix that has multiple levels of wrongness and behaves erraticly ... but didn't have time to fix it,  
In a nutshell, it makes a copy of each `pv.DataSet` registered in the `pv.Plotter` (from the `pv.Actor.mapper`) and I have a toolbox shortcut that pops-up a `pv.plane_widget()` that I use to clip all my `pv.DataSet` copies (should run in the background). When the tool is activated, I re-map the actors to display the clipped copy instead of the original DataSet. 

```py
class toolbox(QToolBar):
    def __init__(self, renderer: QtInteractor) -> None:
        super().__init__(renderer)
        self.plotter = renderer
     
#  [...]      
        
    def clip_scene(self, state: bool = True):
        if state:
            self.backup = {
                name: actor.mapper.GetInput().copy()
                for name, actor in self.actors.items()
            }

            def clip_actors(normal, origin):
                origin = np.asarray(origin) / np.asarray(self.plotter.scale)
                for name, actor in self.plotter.actors.items():
                    if name not in self.backup:  # backup any new actor
                        self.backup[name] = actor.mapper.dataset.copy()
                    actor.mapper.dataset = self.backup[name].clip(
                        origin=origin, normal=normal, invert=False
                    )

            self.plotter.add_plane_widget(clip_actors, normal="y", factor=1.0)
        else:
            self.plotter.clear_plane_widgets()
            for name, actor in self.plotter.actors.items():
                actor.mapper.dataset = self.backup.pop(name)
            # cleanup any back
            self.backup = {}
```
